src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.cs

```diff
@@ -0,0 +1,847 @@
+                }
+            }
+        }
+
+        private static void ReportSupportedDiagnostic(IOperation operation, OperationBlockAnalysisContext context, string name, string platformName, string? version = null) =>
+            context.ReportDiagnostic(version == null ? operation.CreateDiagnostic(RequiredOsRule, name, platformName) :
+                operation.CreateDiagnostic(RequiredOsVersionRule, name, platformName, version));
+
+        private static void ReportUnsupportedDiagnostic(IOperation operation, OperationBlockAnalysisContext context, string name, string platformName, string? version = null) =>
+            context.ReportDiagnostic(version == null ? operation.CreateDiagnostic(UnsupportedOsRule, name, platformName) :
+                operation.CreateDiagnostic(UnsupportedOsVersionRule, name, platformName, version));
+
+        private static string? VersionToString(Version version) => IsEmptyVersion(version) ? null : version.ToString();
+
+        private static ISymbol? GetOperationSymbol(IOperation operation)
+            => operation switch
+            {
+                IInvocationOperation iOperation => iOperation.TargetMethod,
+                IObjectCreationOperation cOperation => cOperation.Constructor,
+                IFieldReferenceOperation fOperation => IsWithinConditionalOperation(fOperation) ? null : fOperation.Field,
+                IMemberReferenceOperation mOperation => mOperation.Member,
+                _ => null,
+            };
+
+        private static void AnalyzeOperation(IOperation operation, OperationAnalysisContext context,
+            PooledConcurrentDictionary<IOperation, SmallDictionary<string, PlatformAttributes>> platformSpecificOperations,
+            PooledConcurrentDictionary<ISymbol, SmallDictionary<string, PlatformAttributes>> platformSpecificMembers)
+        {
+            var symbol = GetOperationSymbol(operation);
+
+            if (symbol == null)
+            {
+                return;
+            }
+
+            if (!platformSpecificMembers.TryGetValue(symbol.OriginalDefinition, out var operationAttributes) &&
+                TryFindPlatformAttributesApplied(symbol.GetAttributes(), symbol.OriginalDefinition.ContainingSymbol, out operationAttributes))
+            {
+                platformSpecificMembers.TryAdd(symbol.OriginalDefinition, operationAttributes);
+            }
+
+            if (operationAttributes != null && operationAttributes.Any())
+            {
+                if (!platformSpecificMembers.TryGetValue(context.ContainingSymbol.OriginalDefinition, out var callSiteAttribute) &&
+                    TryFindContainingSymbolPlatformAttributes(context.ContainingSymbol, out callSiteAttribute))
+                {
+                    platformSpecificMembers.TryAdd(context.ContainingSymbol.OriginalDefinition, callSiteAttribute);
+                }
```

Comment:
Same comment as https://github.com/dotnet/roslyn-analyzers/pull/3917/files#r470885248 - we should be adding out our cache even when symbol has no platform attributes, otherwise we will recompute a lot. I think you can actually refactor out the code from both this and above if statement that does similar logic into a local function and share code.